test/restart.jl

```diff
@@ -0,0 +1,335 @@
+# use `let` blocks. When `let` blocks are used, a different TestSet type is used internally,
+# and this type does not add stack traces. The drawback is that it does not collect
+# statistics on how many tests pass and how many fail. Also, when using let blocks, all the
+# nested testsets have to be using let blocks.
+#
+# We need to try-catch because @test thorws exceptions. Exceptions are normally captured by
+# the @testset, but this does not work when using a let block. :(
+#
+# And there's a final snag: by try-catching, the error is not propagated, so the
+# job is not marked as failing. We have to manually trigger exit with an error
+# status when the job has failed.
+FAILED = Ref(0)
+macro testmsg(ex, str)
+    esc(quote
+        try
+            @testset let name = $str
+                @test $ex
+            end
+        catch
+            FAILED[] = 1
+        end
+    end)
+end
+
+# This test checks that:
+
+# 1. A simulation, saved to a checkpoint, is read back identically (up to some
+#   tolerance and excluding those fields that are computed during the
+#   calculation of the tendencies)
+# 2. A simulation, saved to a previous checkpoint, and read back and evolved to
+#   the same time is identical (up to some tolerance)
+# 3. ClimaAtmos can automatically detect restarts
+
+"""
+    _error(arr1::AbstractArray, arr2::AbstractArray; ABS_TOL = 100eps(eltype(arr1)))
+
+We compute the error in this way:
+- when the absolute value is larger than ABS_TOL, we use the absolute error
+- in the other cases, we compare the relative errors
+"""
+function _error(
+    arr1::AbstractArray,
+    arr2::AbstractArray;
+    ABS_TOL = 100eps(eltype(arr1)),
+)
+    diff = abs.(arr1 .- arr2)
+    denominator = abs.(arr1)
+    error = ifelse.(denominator .> ABS_TOL, diff ./ denominator, diff)
+    return error
+end
+
+"""
+    compare(v1, v2; name = "", ignore = Set([:rc]))
+
+Recursively compare `v1` and `v2` up to some numeric tolerance.
+
+`compare` walks through all the properties in `v1` and `v2` until it finds
+that there are no more properties. At that point, `compare` tries to match the
+resulting objects. When such objects are arrays with floating point, `compare`
+defines a notion of `error` that is the following: when the absolute value is
+less than `100eps(eltype)`, `error = absolute_error`, otherwise it is relative
+error. The `error` is then compared against a tolerance.
+
+Keyword arguments
+=================
+
+- `name` is used to collect the name of the property while we go recursively
+  over all the properties. You can pass a base name.
+- `ignore` is a collection of `Symbol`s that identify properties that are
+  ignored when walking through the tree. This is useful for properties that
+  are known to be different (e.g., `output_dir`).
+
+`:rc` is some CUDA/CuArray internal object that we don't care about
+"""
+function compare(v1, v2; name = "", ignore = Set([:rc]))
```

Comment:
Let's chat offline about how we can improve this function, or somehow use `Fields.rcompare`/`Fields.@rprint_diff`, or write a refactored form of this function that satisfies the needs of this test.

Reply:
Some of the concrete issues are:
 - It does too many things (comparison, performs different tests, has exclusions, tracks names)
 - The macro + closure over `FAILED` makes this function not portable
 - There's code duplication in leaf-level implementations
 - The comparisons are inconsistent (some with `==` and others with `_error`)

I think we can refactor this to keep most of the same functionality, but it'd be helpful to discuss this.